### PR TITLE
Fix/full attention ghost block race

### PR DESCRIPTION
**Pull request by jhaotingc**
## Purpose

`FullAttentionManager.cache_blocks()` commits prefix block hashes to the shared `BlockPool` hash table **before** the GPU forward pass writes the KV values.
In the same scheduling step, `get_computed_blocks()` calls `find_longest_cache_hit` for a newly-admitted request and finds those uncommitted blocks as cache hits — the request skips re-prefilling that suffix and attends over zero/stale KV.
`MambaManager` already guards against this with `cached_blocks_this_step` (PR #29387). `FullAttentionManager` had no equivalent guard.

This PR mirrors `MambaManager` exactly:
- `__init__`: add `cached_blocks_this_step: set[BlockHashWithGroupId]`
- `cache_blocks()`: record newly committed block hashes each step
- `new_step_starts()`: clear the set each scheduling step
- `get_num_blocks_to_allocate()`: return `num_gpu_blocks + 1` when the tail of the prefix hit chain is a ghost block, causing the scheduler to defer the request to the next step (when the GPU write is complete)

Ghost blocks always form a contiguous suffix of the hit chain (prefix hash chaining guarantees this), so checking the last block suffices. The deferral is exactly one step — by the next scheduling cycle the GPU has written the KV and the cache hit is legitimate.

"Possible" fix to #41838.

## Test Plan

Serve GPT-OSS-120B on 2xH200 [openai/gpt-oss-120b](https://huggingface.co/openai/gpt-oss-120b) and [nvidia/gpt-oss-120b-Eagle3-v3](https://huggingface.co/nvidia/gpt-oss-120b-Eagle3-v3)

```
vllm serve /trt_llm_ci/data/llm-models/gpt_oss/gpt-oss-120b \
--max-num-seqs 16 \
--max-num-batched-tokens 4096 \
--max-model-len 65536 \
--host 0.0.0.0 \
--port 8001 \
--tensor-parallel-size 2 \
--gpu-memory-utilization 0.9 \
--trust-remote-code \
--async-scheduling \
--enable-prefix-caching \
--max-cudagraph-capture-size 1024 \
--stream-interval 20 \
--speculative-config '{"model":"/trt_llm_ci/data/llm-models/gpt_oss/gpt-oss-120b-Eagle3","num_speculative_tokens":3,"method":"eagle3"}' \
--compilation-config '{"use_inductor_graph_partition":1,"pass_config":{"enable_qk_norm_rope_fusion":true,"eliminate_noops":true,"fuse_allreduce_rms":true}}' \
  2>&1 | tee out.log &
```

Then run send the same request 4 times with concurrency=4. The request has 31408 tokens after tokenized.

## Test Result

Before, the req[1-3] is computed with the same prefill step as req[0]'s last block, causing referring to not-yet-computed "ghost" KV block.
Image

After, req[1-3] is computed **After** the req[0] prefill request is completed, dragging 1 step but necessary for correct computation.
Image

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by orozery** [comment]
@jhaotingc this looks similar to #38715.
Please see discussion over there.

**Comment by jhaotingc** [comment]
I tested kernel-level correctness when requests hit are scheduled with the requests to be computed in the same step. Not able to repro evidence of logits degradation. 

Refining the solution to opt-in basis for #41838, given that this resolved the **Eagle3 AL degrades by time** issue.

cc @orozery if you have any thoughts. Thank you!

**Review by ZJY0516 — COMMENTED** [review]
Could you please provide a test that can trigger this race like the test in https://github.com/vllm-project/vllm/pull/43569
